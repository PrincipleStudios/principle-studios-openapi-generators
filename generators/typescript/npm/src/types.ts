import { RequestOpts, ResponseArgs } from "./inputs-outputs";

export type RequestBodies = {
    [mimeType: string]: any;
}

export type Responses = {
    'other'?: {
        [mimeType: string]: any;
    };
    [statusCode: number]: {
        [mimeType: string]: any;
    }
}

export type StandardResponse<TStatusCode extends number | 'other' = number | 'other', TMimeType extends string = string, TBody extends unknown = unknown> =
    { statusCode: TStatusCode; mimeType: TMimeType; data: TBody; response: ResponseArgs };

export type TransformRequestNoBody<TRequestParams extends {}, TResult> = (params: TRequestParams) => TResult;
export type TransformRequestWithBody<TRequestParams extends {}, TRequestBodies extends RequestBodies, TResult> =
    <K extends keyof TRequestBodies>(params: TRequestParams, body: TRequestBodies[K], mimeType: K) => TResult;
export type TransformRequestWithOptionalBody<TRequestParams extends {}, TRequestBodies extends RequestBodies, TResult> =
    TransformRequestNoBody<TRequestParams, TResult> & TransformRequestWithBody<TRequestParams, TRequestBodies, TResult>;

export type TransformCallType = 'no-body' | 'body' | 'optional';

export type TransformRequest<TRequestParams extends {}, TRequestBodies extends RequestBodies, TCallType extends TransformCallType, TResult> =
    TCallType extends 'no-body' ? TransformRequestNoBody<TRequestParams, TResult>
    : TCallType extends 'body' ? TransformRequestWithBody<TRequestParams, TRequestBodies, TResult>
    : TCallType extends 'optional' ? TransformRequestWithOptionalBody<TRequestParams, TRequestBodies, TResult>
    : never;

export type TransformResponse<TResponses extends StandardResponse> = (args: ResponseArgs) => TResponses;

export type RequestConversion<TRequestParams extends {}, TRequestBodies extends RequestBodies, TResponses extends StandardResponse, TCallType extends TransformCallType> = {
    request: TransformRequest<TRequestParams, TRequestBodies, TCallType, RequestOpts>;
    response: TransformResponse<TResponses>;
};

export type RequestConversions = Record<string, RequestConversion<any, any, StandardResponse, TransformCallType>>;
