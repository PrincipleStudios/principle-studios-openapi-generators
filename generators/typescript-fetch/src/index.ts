import type {
	RequestBodies,
	RequestConversion,
	StandardResponse,
	TransformCallType,
	AdapterRequestArgs,
	RequestConversions,
	HttpMethod,
	AdapterResponseArgs,
} from '@principlestudios/openapi-codegen-typescript';

const applicationJson = 'application/json';

type AnyObject = Record<string, any>;
type EmptyObject = Record<never, never>;
type AnyRequestConversion = RequestConversion<
	HttpMethod,
	AnyObject,
	AnyObject,
	RequestBodies,
	StandardResponse,
	TransformCallType
>;
type IfKeyless<T, TTrue, TFalse> = EmptyObject extends T ? TTrue : TFalse;

type FetchRequest = {
	method: HttpMethod;
	headers: Record<string, string> | undefined;
	body: FormData | string;
};

type FetchResponse = {
	status: number;
	headers: Headers;
	json(): Promise<unknown>;
	body: unknown; // ReadableStream, but may be different for Node vs DOM
};

export type BaseFetchImplementation = (url: string | URL, requestInit: FetchRequest) => Promise<FetchResponse>;
export type FetchImplementation<TExtra> = (
	...params: [...Parameters<BaseFetchImplementation>, TExtra]
) => Promise<FetchResponse>;

function createRequestArgs(requestOpts: AdapterRequestArgs): Parameters<BaseFetchImplementation> {
	const url = requestOpts.path;

	return [
		url,
		{
			method: requestOpts.method,
			headers: requestOpts.headers
				? Object.fromEntries(Object.entries(requestOpts.headers).filter((t): t is [string, string] => t[1] !== null))
				: undefined,
			body:
				requestOpts.headers && requestOpts.headers['Content-Type'] === 'application/x-www-form-urlencoded'
					? (requestOpts.body as FormData)
					: JSON.stringify(requestOpts.body),
		},
	];
}

async function createResponseArgs(
	fetchResponse: Awaited<ReturnType<BaseFetchImplementation>>
): Promise<AdapterResponseArgs> {
	const response = await fetchResponse;
	const contentType = response.headers.get('Content-Type') ?? '';
	return {
		status: response.status,
		response: contentType.split(';')[0] === applicationJson ? await response.json() : response.body,
		getResponseHeader(header) {
			return response.headers.get(header);
		},
	};
}

export function toFetchOperation<
	TParams extends AnyObject,
	TBody extends RequestBodies,
	TResponse extends StandardResponse,
	TCallType extends TransformCallType,
	TExtra
>(
	fetchImpl: FetchImplementation<TExtra>,
	conversion: RequestConversion<HttpMethod, AnyObject, TParams, TBody, TResponse, TCallType>
): Converted<AnyRequestConversion, TExtra> {
	return async function transform(...[param]): Promise<TResponse> {
		const { params = {}, body = undefined, extra } = param ?? {};
		const requestArgs: AdapterRequestArgs = conversion.request(
			params as TParams,
			body as TBody['application/json'],
			(body ? applicationJson : undefined) as keyof TBody
		);
		const fetchArgs = createRequestArgs(requestArgs);
		const fetchResponse = await fetchImpl(...fetchArgs, extra as TExtra);
		const adapterResponseArgs = await createResponseArgs(fetchResponse);
		return conversion.response(adapterResponseArgs);
	};
}

type ParamPart<TParams> = IfKeyless<TParams, { params?: TParams }, { params: TParams }>;
type NoBody = { body?: undefined };
type BodyPartInner<TBodies extends RequestBodies> = { body: TBodies['application/json'] };
type BodyPart<TBodies extends RequestBodies, TCallType extends TransformCallType> = TCallType extends 'no-body'
	? NoBody
	: TCallType extends 'optional'
	? BodyPartInner<TBodies> | NoBody
	: BodyPartInner<TBodies>;
type ExtraPart<TExtra> = undefined extends TExtra ? { extra?: TExtra } : { extra: TExtra };

type RequestParam<
	TCallType extends TransformCallType,
	TParams,
	TBodies extends RequestBodies,
	TExtra
> = ParamPart<TParams> & BodyPart<TBodies, TCallType> & ExtraPart<TExtra>;

type ConvertedParams<TCallType extends TransformCallType, TParams, TBodies extends RequestBodies, TExtra> = IfKeyless<
	RequestParam<TCallType, TParams, TBodies, TExtra>,
	[req?: RequestParam<TCallType, TParams, TBodies, TExtra>],
	[req: RequestParam<TCallType, TParams, TBodies, TExtra>]
>;

type Converted<TConversion extends AnyRequestConversion, TExtra> = TConversion extends RequestConversion<
	HttpMethod,
	AnyObject,
	infer TParams,
	infer TBodies,
	infer TResponse,
	infer TCallType
>
	? (...args: ConvertedParams<TCallType, TParams, TBodies, TExtra>) => Promise<TResponse>
	: never;

function applyTransform<TMethods extends RequestConversions, TExtra>(
	methods: TMethods,
	transform: (input: AnyRequestConversion) => Converted<AnyRequestConversion, TExtra>
): {
	[K in keyof TMethods]: Converted<TMethods[K], TExtra>;
} {
	return Object.fromEntries(
		Object.entries(methods).map(([operationId, conversion]) => [operationId, transform(conversion)])
	) as {
		[K in keyof TMethods]: Converted<TMethods[K], TExtra>;
	};
}

export function toFetchApi<TMethods extends RequestConversions, TExtra>(
	api: TMethods,
	fetchImpl: FetchImplementation<TExtra>
) {
	return applyTransform<TMethods, TExtra>(api, (conversion) => toFetchOperation(fetchImpl, conversion));
}
