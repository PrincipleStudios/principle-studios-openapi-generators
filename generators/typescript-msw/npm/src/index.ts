/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/naming-convention */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
// TODO: this likely needs to be rewritten to support newer msw anyway
import type {
	AdapterRequestArgs,
	HttpMethod,
	RequestBodies,
	RequestConversion,
	StandardResponse,
	TransformCallType,
} from '@principlestudios/openapi-codegen-typescript';
import type {
	DefaultBodyType,
	Match,
	MockedRequest,
	MockedResponse,
	ResponseResolverReturnType,
	ResponseComposition,
	RestContext,
} from 'msw';
import { RestHandler } from 'msw';

function deepEqual(x: unknown, y: unknown): boolean {
	const tx = typeof x,
		ty = typeof y;
	if (!x || !y || tx !== 'object' || tx !== ty) return x === y;

	const xKeys = Object.keys(x),
		yKeys = Object.keys(y);
	return (
		xKeys.length === yKeys.length &&
		xKeys.every((key) =>
			deepEqual(
				(x as Record<string, unknown>)[key],
				(y as Record<string, unknown>)[key],
			),
		)
	);
}

export type AsyncResponseResolver<
	RequestType = MockedRequest,
	BodyType extends DefaultBodyType = any,
> = (
	req: RequestType,
	res: ResponseComposition<BodyType>,
	context: RestContext,
) =>
	| ResponseResolverReturnType<MockedResponse<BodyType>>
	| Promise<ResponseResolverReturnType<MockedResponse<BodyType>>>;

export type SafeResponse<T extends StandardResponse = StandardResponse> = Omit<
	T,
	'response'
> & { headers?: Record<string, unknown> };

export function toMswResponse(
	response: SafeResponse,
	res: ResponseComposition,
	ctx: RestContext,
) {
	return res(
		ctx.status(response.statusCode === 'other' ? 0 : response.statusCode),
		ctx.json(response.data),
		// TODO: support headers
	);
}

export class MappedRestHandler<
	T extends MockedRequest = MockedRequest,
> extends RestHandler<T> {
	constructor(
		private request: AdapterRequestArgs,
		response: SafeResponse | AsyncResponseResolver<T>,
	) {
		super(
			request.method,
			new URL(request.path.split('?')[0], 'http://localhost').toString(),
			async (
				req: T,
				res,
				ctx,
			): Promise<ResponseResolverReturnType<MockedResponse<any>>> => {
				if ('body' in request) {
					if (!deepEqual(request.body, await req.json())) {
						this.shouldSkip = true;
						return res();
					}
				} else if (req.bodyUsed) {
					this.shouldSkip = true;
					return res();
				}
				// TODO: request headers aren't checked

				if (typeof response === 'function') return response(req, res, ctx);
				return toMswResponse(response, res, ctx);
			},
		);
	}

	override run(
		...params: Parameters<RestHandler<T>['run']>
	): ReturnType<RestHandler<T>['run']> {
		this.shouldSkip = false;
		return super.run(...params);
	}

	override predicate(request: T, parsedResult: Match): boolean {
		// body isn't checked here because this can't be async, so instead it is checked in the MappedRestHandler above
		return (
			super.predicate(request, parsedResult) && this.queryStringMatch(request)
		);
	}

	private queryStringMatch(request: T) {
		return (
			request.url.searchParams.toString() ==
			(this.request.path.split('?')[1] ?? '')
		);
	}
}

export function toMswHandler<
	TMethod extends HttpMethod,
	TUrlParams extends {},
	TRequestParams extends TUrlParams,
	TRequestBodies extends RequestBodies,
	TResponses extends StandardResponse,
	TCallType extends TransformCallType,
>(
	conversion: RequestConversion<
		TMethod,
		TUrlParams,
		TRequestParams,
		TRequestBodies,
		TResponses,
		TCallType
	>,
) {
	type MyRequest =
		| (TCallType extends 'no-body' | 'optional'
				? { params: TRequestParams; body?: never; mimeType?: never }
				: never)
		| (TCallType extends 'body' | 'optional'
				? {
						[K in keyof TRequestBodies]: {
							params: TRequestParams;
							body?: TRequestBodies[K];
							mimeType?: K;
						};
					}[keyof TRequestBodies]
				: never);

	return function (
		request: MyRequest,
		response: AsyncResponseResolver | SafeResponse<TResponses>,
	) {
		const standardRequest: AdapterRequestArgs =
			'body' in request
				? conversion.request(
						request.params,
						request.body as TRequestBodies[keyof TRequestBodies],
						request.mimeType as keyof TRequestBodies,
					)
				: {
						method: conversion.method,
						path: conversion.url(request.params),
						// TODO: headers support?
					};
		return new MappedRestHandler(standardRequest, response);
	};
}
