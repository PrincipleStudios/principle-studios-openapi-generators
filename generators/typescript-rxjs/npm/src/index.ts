import { RequestBodies, RequestConversion, StandardResponse, TransformCallType, TransformRequest, RequestOpts, RequestConversions, HttpQuery, encodeURI } from '@principlestudios/openapi-codegen-typescript';

import { Observable, Subject, Subscriber, merge } from 'rxjs';
import { map } from 'rxjs/operators';
import { ajax, AjaxRequest, AjaxResponse } from 'rxjs/ajax';

const queryString = (params: HttpQuery): string => Object.keys(params)
    .map((key) => {
        const value = params[key];
        return (value instanceof Array)
            ? value.map((val) => `${encodeURI(key)}=${encodeURI(val)}`).join('&')
            : `${encodeURI(key)}=${encodeURI(value)}`;
    })
    .join('&');

function rxWithPrefix(prefix: string, rxjsRequest: (params: AjaxRequest) => Observable<AjaxResponse> = ajax) {

    const createRequestArgs = (requestOpts: RequestOpts): AjaxRequest => {
        let url = prefix + requestOpts.path;
        if (requestOpts.query !== undefined && Object.keys(requestOpts.query).length !== 0) {
            // only add the queryString to the URL if there are query parameters.
            // this is done to avoid urls ending with a '?' character which buggy webservers
            // do not handle correctly sometimes.
            url += '?' + queryString(requestOpts.query);
        }

        return {
            url,
            method: requestOpts.method,
            headers: requestOpts.headers,
            body: (requestOpts.headers && requestOpts.headers['Content-Type'] === 'application/x-www-form-urlencoded') ? requestOpts.body : JSON.stringify(requestOpts.body),
            responseType: requestOpts.responseType || 'json',
        };
    }

    return function fetchRequest<TParams extends {}, TBody extends RequestBodies, TResponse extends StandardResponse, TCallType extends TransformCallType>(
        conversion: RequestConversion<TParams, TBody, TResponse, TCallType>
    ): TransformRequest<TParams, TBody, TCallType, Observable<TResponse>> {

        function transform(params: TParams): Observable<TResponse>;
        function transform<K extends keyof TBody>(params: TParams, body: TBody[K], mimeType: K): Observable<TResponse>;
        function transform(params: TParams, body: TBody[any] | null = null, mimeType: string | null = null): Observable<TResponse> {
            const requestOpts: RequestOpts = conversion.request(params, body!, mimeType!);
            return rxjsRequest(createRequestArgs(requestOpts))
                .pipe(map(response => conversion.response({
                    status: response.status,
                    response: response.response,
                    getResponseHeader(header) { return response.xhr.getResponseHeader(header); },
                })));
        }
        return transform as TransformRequest<TParams, TBody, TCallType, Observable<TResponse>>;
    }
}

type Converted<TConversion extends RequestConversion<any, any, any, any>> = TConversion extends RequestConversion<infer TParams, infer TBody, infer TResponse, infer TCallType> ? TransformRequest<TParams, TBody, TCallType, Observable<TResponse>> : never;

function applyTransform<TMethods extends RequestConversions>(
    methods: TMethods,
    transform: (input: RequestConversion<any, any, any, any>) => Converted<RequestConversion<any, any, any, any>>
): {
        [K in keyof TMethods]: Converted<TMethods[K]>;
    } {
    return Object.keys(methods).reduce((prev, next) => ({ ...prev, [next]: transform(methods[next]) }), {} as {
        [K in keyof TMethods]: Converted<TMethods[K]>;
    });
}

export function toRxjsApi<TMethods extends RequestConversions>(api: TMethods, prefix: string = '', rxjsRequest: (params: AjaxRequest) => Observable<AjaxResponse> = ajax) {
    return applyTransform(api, rxWithPrefix(prefix, rxjsRequest));
}
