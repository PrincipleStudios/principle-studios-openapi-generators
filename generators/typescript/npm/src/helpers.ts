
export const encodeURI = (value: unknown) => encodeURIComponent(String(value));

export const throwIfNullOrUndefined = (value: unknown, nickname?: string) => {
    if (value == null) {
        throw new Error(`Parameter "${value}" was null or undefined when calling "${nickname}".`);
    }
};
