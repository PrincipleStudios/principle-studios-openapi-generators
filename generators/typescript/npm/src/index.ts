export * from './helpers';
export * from './inputs-outputs';
