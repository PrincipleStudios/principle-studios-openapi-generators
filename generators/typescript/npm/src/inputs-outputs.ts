
export type HttpMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE' | 'OPTIONS' | 'HEAD';
export type HttpHeaders = { [key: string]: string | null; };

/** Standardized request to be sent to the adapter layer */
export type AdapterRequestArgs = {
    path: string;
    method: HttpMethod;
    headers?: HttpHeaders;
    body?: unknown;
    responseType?: 'json' | 'blob' | 'arraybuffer' | 'text';
}

/** Standardize response to be received from the adapter layer */
export type AdapterResponseArgs = {
    status: number;
    getResponseHeader(header: string): string | null;
    response: unknown;
};
