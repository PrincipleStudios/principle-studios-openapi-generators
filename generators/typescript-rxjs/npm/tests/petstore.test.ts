import * as path from 'path';
import { toRxjsApi } from '../src';
import { runCodegen } from './runCodegen';
import { verifyAllFiles } from './verifyAllFiles';
import operations from './petstore/operations';
import type {Responses} from './petstore/operations/addPet';
import type { Observable } from 'rxjs';

describe('typescript-rxjs petstore.yaml', () => {
    const outputDir = path.join(__dirname, "../out.local.petstore");
    const source = path.join(__dirname, "../../../../schemas/petstore.yaml");

    beforeAll(async () => {
        await runCodegen(source, outputDir, `-c`);
    })

    verifyAllFiles(outputDir);

    it('can be wrapped with rxjs', () => {
        const wrapped = toRxjsApi(operations);
        const petResponse: Observable<Responses> = wrapped.addPet({}, { name: 'Fido', tag: 'dog' }, 'application/json');
    })
});