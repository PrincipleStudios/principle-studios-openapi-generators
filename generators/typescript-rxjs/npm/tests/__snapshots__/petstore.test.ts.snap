// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typescript-rxjs petstore.yaml can generate the files: C:\\Users\\mattd\\Source\\openapi-codegen\\generators\\typescript-rxjs\\npm\\out.local.petstore\\models\\_Error.ts 1`] = `
"/*
 * Swagger Petstore
 *
 * A sample API that uses a petstore as an example to demonstrate features in the OpenAPI 3.0 specification
 *
 * OpenAPI spec version: 1.0.0
 * Contact: apiteam@swagger.io
 * Generated by: https://principle.tools (PrincipleStudios.OpenApiCodegen.Client.TypeScript v1.0.0.0)
 */

/**
 * @export
 * @interface _Error
 */
export type _Error = {
    /**
     * @type number
     */
    code: number;
    /**
     * @type string
     */
    message: string;
};
"
`;

exports[`typescript-rxjs petstore.yaml can generate the files: C:\\Users\\mattd\\Source\\openapi-codegen\\generators\\typescript-rxjs\\npm\\out.local.petstore\\models\\NewPet.ts 1`] = `
"/*
 * Swagger Petstore
 *
 * A sample API that uses a petstore as an example to demonstrate features in the OpenAPI 3.0 specification
 *
 * OpenAPI spec version: 1.0.0
 * Contact: apiteam@swagger.io
 * Generated by: https://principle.tools (PrincipleStudios.OpenApiCodegen.Client.TypeScript v1.0.0.0)
 */

/**
 * @export
 * @interface NewPet
 */
export type NewPet = {
    /**
     * @type string
     */
    name: string;
    /**
     * @type string | null
     */
    tag?: string | null;
};
"
`;

exports[`typescript-rxjs petstore.yaml can generate the files: C:\\Users\\mattd\\Source\\openapi-codegen\\generators\\typescript-rxjs\\npm\\out.local.petstore\\models\\Pet.ts 1`] = `
"/*
 * Swagger Petstore
 *
 * A sample API that uses a petstore as an example to demonstrate features in the OpenAPI 3.0 specification
 *
 * OpenAPI spec version: 1.0.0
 * Contact: apiteam@swagger.io
 * Generated by: https://principle.tools (PrincipleStudios.OpenApiCodegen.Client.TypeScript v1.0.0.0)
 */

/**
 * @export
 * @interface Pet
 */
export type Pet = {
    /**
     * @type string
     */
    name: string;
    /**
     * @type string | null
     */
    tag?: string | null;
    /**
     * @type BigInt
     */
    id: BigInt;
};
"
`;

exports[`typescript-rxjs petstore.yaml can generate the files: C:\\Users\\mattd\\Source\\openapi-codegen\\generators\\typescript-rxjs\\npm\\out.local.petstore\\operations\\addPet.ts 1`] = `
"/*
 * Swagger Petstore
 *
 * A sample API that uses a petstore as an example to demonstrate features in the OpenAPI 3.0 specification
 *
 * OpenAPI spec version: 1.0.0
 * Contact: apiteam@swagger.io
 * Generated by: https://principle.tools (PrincipleStudios.OpenApiCodegen.Client.TypeScript v1.0.0.0)
 */
import { HttpHeaders, HttpQuery, RequestOpts, ResponseArgs, throwIfNullOrUndefined, encodeURI, RequestConversion, StandardResponse, TransformRequest } from '@principlestudios/openapi-codegen-typescript';
import {
    _Error,
} from './../models/_Error';
import {
    NewPet,
} from './../models/NewPet';
import {
    Pet,
} from './../models/Pet';

export type RequestParams = {
};
export type RequestBodies = {
    'application/json': NewPet;
};

export type StructuredResponses = {
    /** unexpected error */
    'other': {
        'application/json': _Error;
    };
    /** pet response */
    200: {
        'application/json': Pet;
    };
};
export type Responses = 
    | StandardResponse<'other', 'application/json', _Error>
    | StandardResponse<200, 'application/json', Pet>;

export function addPet<K extends keyof RequestBodies>(params: RequestParams, body: RequestBodies[K], mimeType: K): RequestOpts;
export function addPet(params: RequestParams, body: RequestBodies[keyof RequestBodies] | null = null, mimeType: keyof RequestBodies | null = null): RequestOpts {
    throwIfNullOrUndefined(params, 'params');
    throwIfNullOrUndefined(body, 'body');
    throwIfNullOrUndefined(mimeType, 'mimeType');

    const headers: HttpHeaders = {};
    if (mimeType) headers['Content-Type'] = mimeType;

    return {
        path: '/pets',
        method: 'POST',
        headers,
        body,
    };
}

export const responseConversion = (res: ResponseArgs) => ({
        statusCode: [
            200,
        ].indexOf(res.status) === -1 ? 'other' : res.status,
        mimeType: res.getResponseHeader(\\"Content-Type\\"),
        data: res.response,
        response: res,
    }) as Responses;

export const conversion: RequestConversion<RequestParams, RequestBodies, Responses, 'body'> = {
    request: addPet,
    response: responseConversion,
};

"
`;

exports[`typescript-rxjs petstore.yaml can generate the files: C:\\Users\\mattd\\Source\\openapi-codegen\\generators\\typescript-rxjs\\npm\\out.local.petstore\\operations\\deletePet.ts 1`] = `
"/*
 * Swagger Petstore
 *
 * A sample API that uses a petstore as an example to demonstrate features in the OpenAPI 3.0 specification
 *
 * OpenAPI spec version: 1.0.0
 * Contact: apiteam@swagger.io
 * Generated by: https://principle.tools (PrincipleStudios.OpenApiCodegen.Client.TypeScript v1.0.0.0)
 */
import { HttpHeaders, HttpQuery, RequestOpts, ResponseArgs, throwIfNullOrUndefined, encodeURI, RequestConversion, StandardResponse, TransformRequest } from '@principlestudios/openapi-codegen-typescript';
import {
    _Error,
} from './../models/_Error';

export type RequestParams = {
    id: BigInt;
};
export type RequestBodies = {
};

export type StructuredResponses = {
    /** unexpected error */
    'other': {
        'application/json': _Error;
    };
    /** pet deleted */
    204: {
        '': undefined;
    };
};
export type Responses = 
    | StandardResponse<'other', 'application/json', _Error>
    | StandardResponse<204, '', undefined>;

export function deletePet(params: RequestParams): RequestOpts;
export function deletePet(params: RequestParams): RequestOpts {
    throwIfNullOrUndefined(params, 'params');
    throwIfNullOrUndefined(params.id, 'id');

    const headers: HttpHeaders = {};

    return {
        path: '/pets/{id}'.replace('{id}', encodeURI(params.id)),
        method: 'DELETE',
        headers,
    };
}

export const responseConversion = (res: ResponseArgs) => ({
        statusCode: [
            204,
        ].indexOf(res.status) === -1 ? 'other' : res.status,
        mimeType: res.getResponseHeader(\\"Content-Type\\"),
        data: res.response,
        response: res,
    }) as Responses;

export const conversion: RequestConversion<RequestParams, {}, Responses, 'no-body'> = {
    request: deletePet,
    response: responseConversion,
};

"
`;

exports[`typescript-rxjs petstore.yaml can generate the files: C:\\Users\\mattd\\Source\\openapi-codegen\\generators\\typescript-rxjs\\npm\\out.local.petstore\\operations\\findPetById.ts 1`] = `
"/*
 * Swagger Petstore
 *
 * A sample API that uses a petstore as an example to demonstrate features in the OpenAPI 3.0 specification
 *
 * OpenAPI spec version: 1.0.0
 * Contact: apiteam@swagger.io
 * Generated by: https://principle.tools (PrincipleStudios.OpenApiCodegen.Client.TypeScript v1.0.0.0)
 */
import { HttpHeaders, HttpQuery, RequestOpts, ResponseArgs, throwIfNullOrUndefined, encodeURI, RequestConversion, StandardResponse, TransformRequest } from '@principlestudios/openapi-codegen-typescript';
import {
    _Error,
} from './../models/_Error';
import {
    Pet,
} from './../models/Pet';

export type RequestParams = {
    id: BigInt;
};
export type RequestBodies = {
};

export type StructuredResponses = {
    /** unexpected error */
    'other': {
        'application/json': _Error;
    };
    /** pet response */
    200: {
        'application/json': Pet;
    };
};
export type Responses = 
    | StandardResponse<'other', 'application/json', _Error>
    | StandardResponse<200, 'application/json', Pet>;

export function findPetById(params: RequestParams): RequestOpts;
export function findPetById(params: RequestParams): RequestOpts {
    throwIfNullOrUndefined(params, 'params');
    throwIfNullOrUndefined(params.id, 'id');

    const headers: HttpHeaders = {};

    return {
        path: '/pets/{id}'.replace('{id}', encodeURI(params.id)),
        method: 'GET',
        headers,
    };
}

export const responseConversion = (res: ResponseArgs) => ({
        statusCode: [
            200,
        ].indexOf(res.status) === -1 ? 'other' : res.status,
        mimeType: res.getResponseHeader(\\"Content-Type\\"),
        data: res.response,
        response: res,
    }) as Responses;

export const conversion: RequestConversion<RequestParams, {}, Responses, 'no-body'> = {
    request: findPetById,
    response: responseConversion,
};

"
`;

exports[`typescript-rxjs petstore.yaml can generate the files: C:\\Users\\mattd\\Source\\openapi-codegen\\generators\\typescript-rxjs\\npm\\out.local.petstore\\operations\\findPets.ts 1`] = `
"/*
 * Swagger Petstore
 *
 * A sample API that uses a petstore as an example to demonstrate features in the OpenAPI 3.0 specification
 *
 * OpenAPI spec version: 1.0.0
 * Contact: apiteam@swagger.io
 * Generated by: https://principle.tools (PrincipleStudios.OpenApiCodegen.Client.TypeScript v1.0.0.0)
 */
import { HttpHeaders, HttpQuery, RequestOpts, ResponseArgs, throwIfNullOrUndefined, encodeURI, RequestConversion, StandardResponse, TransformRequest } from '@principlestudios/openapi-codegen-typescript';
import {
    _Error,
} from './../models/_Error';
import {
    Pet,
} from './../models/Pet';

export type RequestParams = {
    tags?: Array<string> | null;
    limit?: number | null;
};
export type RequestBodies = {
};

export type StructuredResponses = {
    /** unexpected error */
    'other': {
        'application/json': _Error;
    };
    /** pet response */
    200: {
        'application/json': Array<Pet>;
    };
};
export type Responses = 
    | StandardResponse<'other', 'application/json', _Error>
    | StandardResponse<200, 'application/json', Array<Pet>>;

export function findPets(params: RequestParams): RequestOpts;
export function findPets(params: RequestParams): RequestOpts {
    throwIfNullOrUndefined(params, 'params');

    const headers: HttpHeaders = {};

    const query: HttpQuery = { 
    };
    if (params.tags != null) { query['tags'] = params.tags; }
    if (params.limit != null) { query['limit'] = params.limit; }

    return {
        path: '/pets',
        method: 'GET',
        headers,
        query,
    };
}

export const responseConversion = (res: ResponseArgs) => ({
        statusCode: [
            200,
        ].indexOf(res.status) === -1 ? 'other' : res.status,
        mimeType: res.getResponseHeader(\\"Content-Type\\"),
        data: res.response,
        response: res,
    }) as Responses;

export const conversion: RequestConversion<RequestParams, {}, Responses, 'no-body'> = {
    request: findPets,
    response: responseConversion,
};

"
`;

exports[`typescript-rxjs petstore.yaml can generate the files: C:\\Users\\mattd\\Source\\openapi-codegen\\generators\\typescript-rxjs\\npm\\out.local.petstore\\operations\\index.ts 1`] = `
"/*
 * Swagger Petstore
 *
 * A sample API that uses a petstore as an example to demonstrate features in the OpenAPI 3.0 specification
 *
 * OpenAPI spec version: 1.0.0
 * Contact: apiteam@swagger.io
 * Generated by: https://principle.tools (PrincipleStudios.OpenApiCodegen.Client.TypeScript v1.0.0.0)
 */
import { conversion as findPets } from './findPets';
import { conversion as addPet } from './addPet';
import { conversion as findPetById } from './findPetById';
import { conversion as deletePet } from './deletePet';

export default {
    findPets,
    addPet,
    findPetById,
    deletePet,
} as const;
"
`;
